From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Mon, 21 Oct 2024 19:28:51 +0800
Subject: [PATCH 36/47] exceptions: Recover SYS_IMP_APL_L2C_ERR_STS more
 properly

At least on A7-A11, this allows m1n1 to not going into an infinite
SError loop on 8-bit writes to UART registers.

Suggested-by: fenfenS <104295404+fenfenS@users.noreply.github.com>
Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 src/cpu_regs.h  |  1 +
 src/exception.c | 10 +++++-----
 2 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/src/cpu_regs.h b/src/cpu_regs.h
index 2c552ef7..85dd8ecd 100644
--- a/src/cpu_regs.h
+++ b/src/cpu_regs.h
@@ -507,6 +507,7 @@
 
 #define L2C_ERR_STS_RECURSIVE_FAULT BIT(1)
 #define L2C_ERR_STS_ACCESS_FAULT    BIT(7)
+#define L2C_ERR_STS_PIO_FAULT       BIT(9)
 #define L2C_ERR_STS_ENABLE_W1C      BIT(56)
 
 #define SYS_IMP_APL_L2C_ERR_ADR sys_reg(3, 3, 15, 9, 0)
diff --git a/src/exception.c b/src/exception.c
index 5bedf6f2..ded20c43 100644
--- a/src/exception.c
+++ b/src/exception.c
@@ -228,9 +228,9 @@ void print_regs(u64 *regs, int el12)
     printf("L2C_ERR_STS: 0x%lx\n", sts);
     printf("L2C_ERR_ADR: 0x%lx\n", mrs(SYS_IMP_APL_L2C_ERR_ADR));
     printf("L2C_ERR_INF: 0x%lx\n", mrs(SYS_IMP_APL_L2C_ERR_INF));
-    if (cpu_features->apple_sysregs_unlocked) {
-        msr(SYS_IMP_APL_L2C_ERR_STS, sts);
-    }
+
+    if (cpu_features->apple_sysregs_unlocked)
+        reg_mask(SYS_IMP_APL_L2C_ERR_STS, L2C_ERR_STS_ACCESS_FAULT | L2C_ERR_STS_PIO_FAULT, 0);
 
     if (is_ecore()) {
         printf("E_LSU_ERR_STS: 0x%lx\n", mrs(SYS_IMP_APL_E_LSU_ERR_STS));
@@ -334,8 +334,8 @@ void exc_sync(u64 *regs)
         print_regs(regs, el12);
 
     if (cpu_features->apple_sysregs_unlocked) {
-        u64 l2c_err_sts = mrs(SYS_IMP_APL_L2C_ERR_STS);
-        msr(SYS_IMP_APL_L2C_ERR_STS, l2c_err_sts); // Clear the L2C_ERR flag bits
+        // Clear the L2C_ERR flag bits
+        reg_mask(SYS_IMP_APL_L2C_ERR_STS, L2C_ERR_STS_ACCESS_FAULT | L2C_ERR_STS_PIO_FAULT, 0);
     }
 
     switch (exc_guard & GUARD_TYPE_MASK) {
